hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java

```diff
@@ -385,32 +393,50 @@ public void deleteCache(String cacheNameWithPrefix, String callerUuid, boolean d
     public CacheConfig putCacheConfigIfAbsent(CacheConfig config) {
         // ensure all configs registered in CacheService are not PreJoinCacheConfig's
         CacheConfig cacheConfig = asCacheConfig(config);
-        CacheConfig localConfig = configs.putIfAbsent(cacheConfig.getNameWithPrefix(), cacheConfig);
-        if (localConfig == null) {
-            if (cacheConfig.isStatisticsEnabled()) {
-                setStatisticsEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
-            }
-            if (cacheConfig.isManagementEnabled()) {
-                setManagementEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+        CacheConfigFuture future = new CacheConfigFuture(nodeEngine, cacheConfig);
+        CacheConfigFuture localConfigFuture = configs.putIfAbsent(cacheConfig.getNameWithPrefix(), future);
+        // if the existing cache config future is not yet fully configured, we block here
+        CacheConfig localConfig = localConfigFuture == null ? null : localConfigFuture.join();
+        if (localConfigFuture == null) {
+            try {
+                if (cacheConfig.isStatisticsEnabled()) {
+                    setStatisticsEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+                }
+                if (cacheConfig.isManagementEnabled()) {
+                    setManagementEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+                }
+                logger.info("Added cache config: " + cacheConfig);
+                additionalCacheConfigSetup(null, config);
+                // now it is safe for others to obtain the new cache config
+                future.complete();
+            } catch (RuntimeException e) {
```

Comment:
i think we can catch `Throwable` here, since `wan` config init on ee side can accept pluggable code from user.

Reply:
+1